### PR TITLE
[types-2.0] Add d3-hexbin

### PR DESCRIPTION
**Pull request by heavenshell**
case 1. Add a new type definition.
- [X] checked compilation succeeds with `--target es6` and `--noImplicitAny` options.
- [X] has correct [naming convention](http://definitelytyped.org/guides/contributing.html#naming-the-file)
- [X] has a [test file](http://definitelytyped.org/guides/contributing.html#tests) with the suffix of  `-tests.ts` or `-tests.tsx`.

`d3-hexbin` is not exists. So we add type definition.

### CODE REVIEW

**Review by tomwanzek — COMMENTED** [review]
@heavenshell Thanks a lot for contributing this definition to the D3 universe. :smile:

Hope you are o.k. with a few changes to consider. They in essence suggest using the same pattern we have applied when writing the other definitions for D3. I.p. to use generics to control type-safety without becoming too restrictive on the one hand or overly permissive by simply using `any`.

This way, we keep developers on a consistent experience.

**Comment by heavenshell** [comment]
@tomwanzek 
Thank you for reviewed my PR!
I'm not familiar with TS but I'll try to update!

**Comment by tomwanzek** [comment]
@heavenshell no worries, if you take a look at the interface for `Quadtree` in **d3-quadtree**, it has a very similar shape compared to how `Hexbin` could be done. The respective definition lines are [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/types-2.0/d3-quadtree/index.d.ts#L25).

I.e. it defines a generator interface with accessors for x- and y-coordinates. And it has a few overloaded factories to create a generator. There are some more than you would need, as the factory can take arguments.

If you have any other questions, don't hesitate. Cheers. T

**Comment by heavenshell** [comment]
@tomwanzek 
Hi, sorry for late response.
I adopted your review point to `d3-hexbin.d.ts`. 
Some tests were failed to build.
https://gist.github.com/heavenshell/748f94b4171cd44e8774e85685f09715

These tests are ported from https://github.com/d3/d3-hexbin/blob/master/test/hexbin-test.js
Failed at `x()` and `y()`.
IMO d3-hexbin.d.ts should pass this tests.

I'm really appreciated if you give me advices to help solove problems.
Thank you.

P.S.
I want to use `d3-hexbin.d.ts` like following.
https://gist.github.com/heavenshell/171f32b8fd3de01ace8a413513d42864

**Comment by tomwanzek** [comment]
@heavenshell With a bit of luck, I can fit it in some time later today. Latest on the weekend.

Could you push a commit with the proposed changes to the definitions file to this PR. This should make it easier to use the review feature to comment on specific lines, if need be.

As for tests, bear mind that one of the design choices of DefinitelyTyped was that, tests are essentially shape tests only. I.e. they only validate that the test file and its related definitions compile without error, they do not emit or execute anything beyond that. So testing the shape is essentially typing the left and right-side of an  assignment independently and checking that the assignment does not fail, same with arguments to functions/methods and their return values.

**Comment by heavenshell** [comment]
@tomwanzek 
Pushed!

>they only validate that the test file and its related definitions compile without error, they do not emit or execute anything beyond that.

Yeah, right.
But I'm worried about coverage of d3-hexbin's spec(use-cases).
I think original tests are covered d3-hexbin use-cases.
So I just want compile similar tests(not a result of test, just a compile test file).
If test file are ok to compile, I think d3-hexbin.d.ts coverage use-cases too.
I think this means user can use d3-hexbin.d.ts safely because test file can compile.

Sorry I'm not a good at English. If you were confused it's so sorry 😢

**Comment by tomwanzek** [comment]
Thanks for pushing the changes, I will look at it later. As for language, no concerns there... my Japanese is distinctly worse. Which is to say, non-existent. Now that is really sad :wink:

**Comment by tomwanzek** [comment]
I had a look at the definitions, and structurally they appear accurate. So it's really just down to completing the tests.

I am copying in a short snippet here that should shed some light on how to apply the definitions in the tests. Clearly, when writing "production code", some of the explicit typing of variables below might be just inferred from an assignment. However, this allows testing the shape using a compile only test.

The key to using the factory, the generator and the accessors, is illustrated by the use of the `Point` interface below, to specify the generic `T`. Note that, I tend to test chainability of setters by explicit reassignment for type-preservation checking, e.g. `hb = hb.x(x);`. Again, when writing actual code, one would simply chain, but here it is about testing that the return type is unchanged for an individual accessor. Otherwise, a return type could have been accidentally changed without failing compilation in the next method invoked in a chain. Hope this clarifies:

```ts
import * as d3Hexbin from 'd3-hexbin';

interface Point {
    x0: number;
    y0: number;
}

let hb: d3Hexbin.Hexbin<Point>;
let bins: d3Hexbin.HexbinBin<Point>[];

// Create generator =======================================

hb = d3Hexbin.hexbin<Point>();

// Configure generator =====================================

// x Accessor ----------------------------------------------

let x: (d:Point) => number;
x = function (d: Point) { return d.x0; };

// test setter
hb = hb.x(x);

// test getter
x = hb.x();

// y Accessor ----------------------------------------------

let y: (d:Point) => number;
y = function (d: Point) { return d.y0; };

// test setter
hb = hb.y(y);

// test getter
y = hb.y();

// Use generator ============================================

bins = hb([
    { x0: 0, y0: 0 }, { x0: 0, y0: 1 }, { x0: 0, y0: 2 },
    { x0: 1, y0: 0 }, { x0: 1, y0: 1 }, { x0: 1, y0: 2 },
    { x0: 2, y0: 0 }, { x0: 2, y0: 1 }, { x0: 2, y0: 2 }
]);

let remappedBins: Array<{
    binCoordinates: [number, number];
    points: Point[];
}>;

bins.map(bin => {
    const x: number = bin.x; // x-coordinate of bin
    const y: number = bin.y; // y-coordinate of bin
    const pointsInBin: Point[] = bin.map(p =>
    {
        const point: Point = p;
        return point;
    });
    return {
        binCoordinates: [x, y],
        points: pointsInBin
    };
});
```

**Comment by heavenshell** [comment]
@tomwanzek 
Thank you for suggesting me a nice snippet.
I'm bit busy but I'll look into it 😄

**Comment by tomwanzek** [comment]
Hey, whenever you're ready. Great that you picked up contributing d3-hexbin! T

**Comment by heavenshell** [comment]
@tomwanzek 
Hi, I updated tests. 
Could you review it?

**Comment by tomwanzek** [comment]
Not sure, if I get to it today. But I will try asap. Thx, T

**Comment by herkulano** [comment]
@tomwanzek @heavenshell 
Hi, I've created a d3-plugin sometime ago and after the upgrade to v4 module I created the `index.d.ts` with the definitions for typescript, similar to the ones you have for `d3-hexbin`:
https://github.com/herkulano/d3-hist2d/blob/master/index.d.ts

I really don't like the structure, it doesn't feel right that in typescript I need to use it as `hist2d() or h2.hist2d()` and in javascript I can use `d3.hist2d()`. 

It feels a lot better to override or merge the d3 namespace, but I don't know how. I tried a few approaches without success. Have you tried this approach also? Or you think that what you have defined is the best approach?

**Comment by tomwanzek** [comment]
@heavenshell I have been working away on moving the initial batches of "core" D3 module definitions to support `strictNullChecks`, which was a todo for a sunny day :wink: As soon as I cleared them up, I will get back to this PR as well.

I had not checked this one for `strictNullChecks` and some of the new linting rules as well. So should be back on it shortly.

@herkulano Is your question on how to use it related to
(1) using your plug-in with a module loader (i.e. using `import`) or
(2) as a plain vanilla script which extends the `d3` global?

**Comment by herkulano** [comment]
@tomwanzek thanks for coming back to me.

My question is not on how to use it, but rather on how to create the `index.d.ts` file so that the plugin can be used in a more "natural" d3 way, like any other d3 module, e.g., `d3.color()`.

It's clear to me how it can be used in "(1) using your plug-in with a module loader (i.e. using import)". In the case of the `d3-hist2d` plugin/module, which also applies to `d3-hexbin` it's:
```js
import { hist2d } from 'd3-hist2d';
hist2d().bins(10)
// OR
import * as h2 from 'd3-hist2d';
h2.hist2d().bins(10)
```

My point is that this feels wrong. With d3 v4 all parts of d3 are modules now, so third-party modules should be the same as d3 default modules, this is true in Javascript so it should also be true in Typescript. It's weird having to change the syntax to use it.

In the definition files we should override/merge the `d3 namespace` instead to achieve this, but unfortunately, I wasn't able to make it work by declaring the d3 namespace in the definition file.

So then overriding/merging the d3 namespace we could then use any plugin like this, or something similar to this:
```js
import 'd3-hist2d';
d3.hist2d().bins(10)
```

**Comment by tomwanzek** [comment]
There is a simple reason, why I asked the question regarding how you would like to use the plug-in.

**Scenario 1: As import with module loader**

> My point is that this feels wrong. With d3 v4 all parts of d3 are modules now, so third-party modules should be the same as d3 default modules, this is true in Javascript so it should also be true in Typescript. It's weird having to change the syntax to use it.

When used as proper modules, all D3 modules are anonymous, so your module definitions already correspond to the way all other definitions are written. There is no magic glue `d3` object that ties them all together when imported, unless you create one using a barrel. This approach is really down to the user of D3 and your plug-in and is simply the way D3 is structured now:

```ts
// d3-with-hist2d-bundle.ts

// re-export all the  "standard" d3-modules you require
// or simply use export * from 'd3' for the entire standard D3 bundle

export * from 'd3-array';
export * from 'd3-selection';
...
export * from 'd3-hist2d';
```
Then if you want to use it as a more "conventional" d3 object simply import from this barrel:

```ts
// whatever-module-imports-d3.ts
import * as d3 from './d3-with-hist2d-bundle';

let x = d3.hist2d(); // et voila
```

**Scenario 2: As plain vanilla script**

This is where you actually run into a typescript vs D3 restriction. D3 code was written in a way, that the created UMD modules _all_ merge into the same `d3` **global** _when used as plain vanilla scripts_.

Assuming, you are writing your module the same way and using Rollup the same way, your JS code will support the same behavior as the other D3 scripts.

Important aside: As a word of caution, avoid clashing with variables/methods/interface names already defined for d3 modules (or their corresponding definition files w.r.t. TypeScript.)

In principle, creating a definition file for a UMD module, can be accomplished by writing the definitions for the proper module and adding:
```ts
export as namespace d3;
```
The definition file for the standard D3 bundle actually does that.

The definition files for the individual modules **do not**. This is due to [TypeScript Issue 9681](https://github.com/Microsoft/TypeScript/issues/9681). The full history of this discussion is also embedded in #9936.

The resolution to this problem can only be achieved with a workaround at the level of the project using the definitions.

Add a directory, e.g.  `typings` in your project, within that directory create a sub-directory named `d3`.

In the `d3` directory, instead of creating a barrel module as described in Scenario 1, create a _definition_ file `index.d.ts` with the following content:

```ts
// index.d.ts

export as namespace d3;

// re-export all the  "standard" d3-modules you require

export * from 'd3-array';
export * from 'd3-selection';
...
export * from 'd3-hist2d';
```
Make sure you add the `typings` directory to the `typeRoots` configuration of your compilation context `tsconfig.json`.

This way, should you write a plain vanilla script in typescript, you should have access to the merged `d3` global by including the compiler directive `/// <reference types='d3' />` at the top of the script.

**Comment by herkulano** [comment]
@tomwanzek thank you so much for your patience and thorough answer, really appreciate it! 💯

**Comment by tomwanzek** [comment]
@herkulano You're welcome, one more module in the ecosystem due to your work, so props to you!

File: d3-hexbin/d3-hexbin-test.ts [hunk 1]
```diff
@@ -0,0 +1,211 @@
+/**
```
Comment (tomwanzek) [inline d3-hexbin/d3-hexbin-test.ts]: As a general note, since the tests in DT are strictly "shape tests", there are a few patterns I tend to follow in the other D3 module definitions:
* When defining variables/constants always explicitly type them with the _expected_ type. I.e. `let x : number = ...` rather than `let x = ...` This way, tests will fail if the right-hand  side does not meet the expectation. Otherwise, the definitions could appear internally consistent due to TypeScripts implicit type inference. The same holds for accessor functions to be tested as the return type of a getter.
* Test each signature for each interface method
* For a "setter" method signature always test the return type individually. Do not use consecutive chaining in shape tests. Otherwise, the test may not find edge cases where a return type is (accidentally) changed by one method, but the next chained method is also defined on it, e.g. a sub-type. Re-assigning to the object on which a setter was invoked is a quick check for chainability.
* As in the case of `hexbin()`, there are sometimes generator factories, which make a default assumption about underlying data types (like each point  is of type `[number, number]`), but allow for more general data structures. In this case by setting the generic `T` e.g. to type `Point`. Again, I would first define a variable with explicit type, like `Hexbin<[number, number]>` or `Hexbin<Point`>, then test the factory by assigning to those variables.
* When there are accessors involved, like here for `x` and `y`, I tend to test against each of the typed generators I obtained from the factory.

One aspect, that is helpful is creating explicit tests for expected failure modes, e.g. assigning `let x: d3Hexbin.Hexbin<Point> = d3Hexbin.hexbin();`. Same for e.g. method signatures with accessor functions depending on a generic. The benefit is that it may catch some edge cases. Note that such tests can only be inspected by you when updating tests. As you should see compiler errors in your IDE right away. You will need to comment them out before committing, so the CI tests do not fail. I tend to do that by also appending a comment like `// fails, argument type mismatch` or something similar.

You can find examples of the latter e.g. in d3-selection.

**Review by tomwanzek — COMMENTED** [review]
I have added a few specific comments and a broad general comment at the beginning of the test file to explain some reasoning for practices I adopted elsewhere.

Hope this helps. Apologies for the slight delay.

**Comment by andy-ms** [comment]
@heavenshell could you respond to the review comments?

**Comment by andy-ms** [comment]
@tomwanzek Merging this, you can make the changes you mentioned if you want to.

**Comment by heavenshell** [comment]
sorry for late response.
I'm glad to fix review point, sorry for my misunderstanding review point.
I think d.ts file would be fine.

**Comment by andy-ms** [comment]
@heavenshell Great! Please do that in a new pull request, though, and based on the `master` branch.

**Comment by heavenshell** [comment]
@andy-ms 
Thx, I still don't understand few review points, but I'll try it.
It's very appriceated if @tomwanzek helps me 👍

File: d3-hexbin/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
+// Type definitions for D3JS d3-hexbin module v0.2.1
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: OK, I'll fix it.

File: d3-hexbin/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+// Project: https://github.com/d3/d3-hexbin/
+// Definitions by: UNCOVER TRUTH Inc. <https://github.com/uncovertruth/>, Tom Wanzek <https://github.com/tomwanzek>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export interface HexbinBin<T> extends Array<T> {
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: I'll fix it

File: d3-hexbin/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    x: number;
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: OK, I'll fix it.

File: d3-hexbin/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    y: number;
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: OK, I'll fix it.

File: d3-hexbin/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+export interface Hexbin<T> {
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: OK, I'll fix it.

File: d3-hexbin/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    /**
+     * Bins the specified array of points, returning an array of hexagonal bins.
+     * For each point in the specified points array, the x- and y-accessors are
+     * invoked to compute the x- and y-coordinates of the point, which is then
+     * used to determine which hexagonal bin to add the point.
+     * If either the x- or y-coordinate is NaN, the point is ignored and will
+     * not be in any of the returned bins.
+     */
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: OK, I'll fix it.

File: d3-hexbin/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * Returns the SVG path string for the hexagon centered at the origin ⟨0,0⟩.
+     * The path string is defined with relative coordinates such that you can
+     * easily translate the hexagon to the desired position.
+     * If radius is not specified, the hexbin’s current radius is used.
+     * If radius is specified, a hexagon with the specified radius is returned;
+     * this is useful for area-encoded bivariate hexbins.
+     *
+     * @param {number} radius Radius number
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: OK, remove `{number}`.
`@param radis An optional radius parameter to be used when generating the hexagon`
is ok?

File: d3-hexbin/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * If x is specified, sets the x-coordinate accessor to the specified
+     * function and returns this hexbin generator.
+     *
+     * The x-coordinate accessor is used by hexbin to compute the x-coordinate
+     * of each point. The default value assumes each point is specified as
+     * a two-element array of numbers [x, y].
+     */
+    x(x: (d: T) => number): Hexbin<T>;
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: What I should write param description?
It's be great if you give me some advice.

File: d3-hexbin/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * If y is specified, sets the y-coordinate accessor to the specified
+     * function and returns this hexbin generator.
+     *
+     * The y-coordinate accessor is used by hexbin to compute the y-coordinate
+     * of each point. The default value assumes each point is specified as
+     * a two-element array of numbers [x, y].
+     */
+    y(y: (d: T) => number): Hexbin<T>
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: same as `x`

File: d3-hexbin/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+     * The extent defaults to [[0, 0], [1, 1]].
+     */
+    extent(): [[number, number], [number, number]];
+}
+
+/**
+ * Constructs a new default hexbin generator.
+ */
+export function hexbin(): Hexbin<[number, number]>;
+export function hexbin<T>(): Hexbin<T>;
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: I'll add JSDoc comment.

File: d3-hexbin/d3-hexbin-test.ts [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+{
+    // hexbin.hexagon(radius) observes the specified radius
+    const path: string = d3Hexbin.hexbin().hexagon(2);
+}
+
+{
+    // hexbin.hexagon(radius) uses the current bin radius if radius is null
+    let path: string = d3Hexbin.hexbin().hexagon(null);
```
Comment (heavenshell) [inline d3-hexbin/d3-hexbin-test.ts]: fmm, so do I need to remove this test?

File: d3-hexbin/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,138 @@
+// Type definitions for D3JS d3-hexbin module v0.2.1
+// Project: https://github.com/d3/d3-hexbin/
+// Definitions by: UNCOVER TRUTH Inc. <https://github.com/uncovertruth/>, Tom Wanzek <https://github.com/tomwanzek>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: OK, I'll add your comment.
```
// Last module patch version validated against: 0.2.1
```

File: d3-hexbin/index.d.ts [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+     * the width of each hexagon is radius × 2 × sin(π / 3)
+     * and the height of each hexagon radius × 3 / 2.
+     */
+    radius(): number;
+
+    /**
+     * If extent is specified, sets the hexbin generator’s extent to the
+     * specified bounds [[x0, y0], [x1, y1]] and returns the hexbin generator.
+     */
+    extent(extent: [[number, number], [number, number]]): Hexbin<T>;
```
Comment (heavenshell) [inline d3-hexbin/index.d.ts]: `@params extent Sets the hexbin generator’s extent`
is ok?

File: d3-hexbin/tsconfig.json [hunk 15]
```diff
@@ -0,0 +1,19 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "es6",
+    "noImplicitAny": true,
+    "strictNullChecks": false,
```
Comment (heavenshell) [inline d3-hexbin/tsconfig.json]: OK, I'll update to `strictNullChecks: true`

**Comment by tomwanzek** [comment]
Will do in a spare minute. Thx for the contribution and merge.
